### DISCUSSION TITLE
What is the best architecture for client implementation?

### DISCUSSION THREAD

**Discussion by Leocete** [panva/openid-client#323] [Q&A] [answered]
Hi, thank you very much for your package - we're using it a lot on our projects and are very satisfied with what possibilities it gives to the developers. 

I'm thinking about the best way of openid-client implementation. What is the best approach to creating a client? To use singleton and call Issuer/create a client only once on app start or re-creating the client every time user makes a call to login?

OAuth flow I'm talking about - Authorization flow.
I'm talking about that part of openid-client implementation:

```
const { Issuer } = require('openid-client');
Issuer.discover('https://accounts.google.com') // => Promise
  .then(function (googleIssuer) {
    console.log('Discovered issuer %s %O', googleIssuer.issuer, googleIssuer.metadata);
  });
  
const client = new googleIssuer.Client({
  client_id: 'zELcpfANLqY7Oqas',
  client_secret: 'TQV5U29k1gHibH5bx1layBo0OSAvAbRT3UYW3EWrSYBB5swxjVfWUa1BS8lqzxG/0v9wruMcrGadany3',
  redirect_uris: ['http://localhost:3000/cb'],
  response_types: ['code'],
  // id_token_signed_response_alg (default "RS256")
  // token_endpoint_auth_method (default "client_secret_basic")
}); // => Client
```

**Comment by panva** [comment, marked as answer]
Discovery is expensive, do it once unless something signals to you a rediscovery is needed. If the client stays the same, makes no sense to reinstantiate.
